File: kerykeion/aspects/aspects_utils.py
# -*- coding: utf-8 -*-
"""
This is part of Kerykeion (C) 2025 Giacomo Battaglia
"""
# TODO: Better documentation and unit tests

from swisseph import difdeg2n
from typing import Union
from kerykeion.schemas.kr_models import AstrologicalSubjectModel, CompositeSubjectModel, PlanetReturnModel
from kerykeion.schemas.kr_literals import AspectMovementType
from kerykeion.schemas.settings_models import KerykeionSettingsCelestialPointModel
from kerykeion.settings.chart_defaults import DEFAULT_CELESTIAL_POINTS_SETTINGS, _CelestialPointSetting


def get_aspect_from_two_points(
    aspects_settings: Union[list[dict], list[dict]],
    point_one: Union[float, int],
    point_two: Union[float, int],
):
    """
    Utility function to calculate the aspects between two points.

    Args:
        aspects_settings (dict): Dictionary containing aspect settings.
        point_one (Union[float, int]): First point.
        point_two (Union[float, int]): Second point.

    Returns:
        dict: Dictionary containing the aspect details.
    """
    distance = abs(difdeg2n(point_one, point_two))
    diff = abs(point_one - point_two)

    for aid, aspect in enumerate(aspects_settings):
        # TODO: Remove the "degree" element EVERYWHERE!
        aspect_degree = aspect["degree"]  # type: ignore
        aspect_orb = aspect["orb"]  # type: ignore

        if (aspect_degree - aspect_orb) <= distance <= (aspect_degree + aspect_orb):
            name = aspect["name"]  # type: ignore
            aspect_degrees = aspect_degree
            verdict = True
            break
    else:
        verdict = False
        name = None
        aspect_degrees = 0

    return {
        "verdict": verdict,
        "name": name,
        "orbit": abs(distance - aspect_degrees),
        "distance": abs(distance - aspect_degrees),
        "aspect_degrees": aspect_degrees,
        "diff": diff,
    }


def calculate_aspect_movement(
    point_one_abs_pos: float,
    point_two_abs_pos: float,
    aspect_degrees: float,
    point_one_speed: float,
    point_two_speed: float,
) -> AspectMovementType:
    """
    Determine whether the aspect orb is decreasing (Applying), increasing
    (Separating), or not changing (Static).

    This implementation uses a "lookahead" approach:
    1. Calculate the current orb (deviation from exact aspect).
    2. Project positions forward by a small time step (dt).
    3. Calculate the future orb.
    4. Compare: if future_orb < current_orb => Applying, else Separating.

    The function handles edge cases including:
    - Retrograde motion (negative speeds)
    - Boundary crossings at 0°/360°
    - Aspects > 180° (normalized correctly)
    - Extreme speed differences (e.g., Ascendant vs slow planets)
    - Very small orbs and speed differences

    Args:
        point_one_abs_pos (float): Absolute position of point 1 (0-360).
        point_two_abs_pos (float): Absolute position of point 2 (0-360).
        aspect_degrees (float): The exact aspect angle (e.g., 0, 60, 90, 120, 180).
            Can be > 180° (will be normalized).
        point_one_speed (float): Speed of point 1 in degrees/day (negative for retrograde).
        point_two_speed (float): Speed of point 2 in degrees/day (negative for retrograde).

    Returns:
        AspectMovementType: "Applying", "Separating", or "Static".

    Raises:
        ValueError: If speed values are None or if positions/aspect are invalid.

    Notes:
        - Static is returned when relative speed is effectively zero (< EPSILON)
        - The function uses a small time step (dt) for numerical differentiation
        - Precision: Changes in orb < EPSILON are considered Static
    """
    # Constants for numerical precision
    # EPSILON for speed comparison: very small relative speeds are considered static
    SPEED_EPSILON = 1e-9
    # EPSILON for orb change comparison: very small orb changes are considered static
    ORB_EPSILON = 1e-6
    # Time step for lookahead (0.001 days ≈ 1.44 minutes)
    # Small enough for linear approximation, large enough to avoid precision issues
    DT = 0.001

    # Validation
    if point_one_speed is None or point_two_speed is None:
        raise ValueError(
            "Speed values for both points are required to compute aspect "
            "movement correctly. point_one_speed and point_two_speed "
            "cannot be None."
        )

    # Validate positions are in valid range
    if not (0 <= point_one_abs_pos < 360) or not (0 <= point_two_abs_pos < 360):
        raise ValueError(f"Positions must be in range [0, 360). Got p1={point_one_abs_pos}, p2={point_two_abs_pos}")

    # Validate aspect degrees
    if aspect_degrees < 0:
        raise ValueError(f"Aspect degrees must be non-negative. Got {aspect_degrees}")

    # If relative speed is effectively zero, the aspect is static.
    # This check prevents division issues and handles the mathematical reality
    # that if both bodies move at the same speed, the orb remains constant.
    relative_speed = abs(point_one_speed - point_two_speed)
    if relative_speed < SPEED_EPSILON:
        return "Static"

    # Helper to calculate the orb (distance from exact aspect)
    def get_orb(p1: float, p2: float, aspect: float) -> float:
        """
        Calculate the orb (deviation from exact aspect).

        Args:
            p1: Position of first point (0-360)
            p2: Position of second point (0-360)
            aspect: Normalized aspect angle (0-180)

        Returns:
            Absolute orb in degrees
        """
        # Calculate shortest distance between points on the circle
        diff = abs(difdeg2n(p1, p2))
        # The orb is the absolute difference between the actual separation and the aspect angle
        return abs(diff - aspect)

    # Normalize aspect to [0, 360) and then to [0, 180]
    # This is necessary because difdeg2n returns the shortest distance (<= 180)
    # For example, aspect=240° is equivalent to 360°-240°=120° on the circle
    aspect_norm = aspect_degrees % 360.0
    if aspect_norm > 180.0:
        aspect_norm = 360.0 - aspect_norm

    # 1. Current state
    current_orb = get_orb(point_one_abs_pos, point_two_abs_pos, aspect_norm)

    # 2. Future state (lookahead)
    # Project positions forward by a small time step
    # Use modulo to handle crossing 0°/360° boundary correctly
    p1_future = (point_one_abs_pos + point_one_speed * DT) % 360.0
    p2_future = (point_two_abs_pos + point_two_speed * DT) % 360.0

    future_orb = get_orb(p1_future, p2_future, aspect_norm)

    # 3. Compare with numerical precision tolerance
    # Calculate the change in orb
    orb_change = future_orb - current_orb

    # Use epsilon for comparison to handle floating point precision issues
    # If the change is smaller than our precision threshold, consider it static
    if abs(orb_change) < ORB_EPSILON:
        return "Static"
    elif orb_change < 0:
        # Orb is decreasing -> bodies are approaching exact aspect
        return "Applying"
    else:
        # Orb is increasing -> bodies are separating from exact aspect
        return "Separating"


def planet_id_decoder(planets_settings: list[KerykeionSettingsCelestialPointModel], name: str) -> int:
    """
    Check if the name of the planet is the same in the settings and return
    the correct id for the planet.
    """
    str_name = str(name)
    for planet in planets_settings:
        if planet["name"] == str_name:
            result = planet["id"]
            return result

    raise ValueError(f"Planet {name} not found in the settings")


def get_active_points_list(
    subject: Union[AstrologicalSubjectModel, CompositeSubjectModel, PlanetReturnModel],
    active_points: list = [],
    *,
    celestial_points: Union[list[_CelestialPointSetting], list[dict]] = DEFAULT_CELESTIAL_POINTS_SETTINGS,
) -> list:
    """
    Given an astrological subject and the settings, return a list of the active points.
    Args:
        subject (AstrologicalSubject): The astrological subject to get the active points from.
        settings (Union[KerykeionSettingsModel, dict]): Settings model o dictionary.

    Returns:
        list: List of the active points.
    """
    point_list = []
    for planet in celestial_points:
        if planet["name"] in active_points:
            point_list.append(subject[planet["name"].lower()])

    return point_list

File: dart/lib/src/aspects/aspects_utils.dart
import 'package:kerykeion_dart/src/types.dart';
import 'package:kerykeion_dart/src/models/astrological_subject.dart';
import 'package:kerykeion_dart/src/settings/chart_defaults.dart';
import 'package:kerykeion_dart/src/models/kerykeion_point.dart';

/// Utility class for aspect calculations
class AspectsUtils {
  /// Calculate the shortest distance between two points on a circle.
  /// Equivalent to Swiss Ephemeris difdeg2n.
  static double difdeg2n(double p1, double p2) {
    double diff = p1 - p2;
    while (diff <= -180.0) {
      diff += 360.0;
    }
    while (diff > 180.0) {
      diff -= 360.0;
    }
    return diff;
  }

  /// Utility function to calculate the aspects between two points.
  static Map<String, dynamic> getAspectFromTwoPoints(List<ChartAspectSetting> aspectsSettings, double pointOne, double pointTwo) {
    final distance = difdeg2n(pointOne, pointTwo).abs();
    final diff = (pointOne - pointTwo).abs();

    bool verdict = false;
    String name = "";
    double aspectDegrees = 0.0;
    double orbit = 0.0;

    for (var aspect in aspectsSettings) {
      final aspectDegree = aspect.degree.toDouble();
      final aspectOrb = aspect.orb?.toDouble() ?? 0.0;

      if ((aspectDegree - aspectOrb) <= distance && distance <= (aspectDegree + aspectOrb)) {
        name = aspect.name;
        aspectDegrees = aspectDegree;
        verdict = true;
        orbit = (distance - aspectDegrees).abs();
        break;
      }
    }

    return {"verdict": verdict, "name": name, "orbit": orbit, "distance": distance, "aspect_degrees": aspectDegrees, "diff": diff};
  }

  /// Determine whether the aspect orb is decreasing (Applying), increasing (Separating), or Static.
  static AspectMovementType calculateAspectMovement(
    double pointOneAbsPos,
    double pointTwoAbsPos,
    double aspectDegrees,
    double pointOneSpeed,
    double pointTwoSpeed,
  ) {
    // Constants for numerical precision
    const double speedEpsilon = 1e-9;
    const double orbEpsilon = 1e-6;
    const double dt = 0.001; // Time step for lookahead (approx 1.44 mins)

    // Validate positions
    if (pointOneAbsPos < 0 || pointOneAbsPos >= 360 || pointTwoAbsPos < 0 || pointTwoAbsPos >= 360) {
      // In Dart we might want to throw or handle gracefully.
      // For now, let's assume inputs are normalized.
      // If strict compliance: throw ArgumentError("Positions must be in range [0, 360)");
    }

    final relativeSpeed = (pointOneSpeed - pointTwoSpeed).abs();
    if (relativeSpeed < speedEpsilon) {
      return AspectMovementType.Static;
    }

    double getOrb(double p1, double p2, double aspect) {
      final diff = difdeg2n(p1, p2).abs();
      return (diff - aspect).abs();
    }

    // Normalize aspect
    double aspectNorm = aspectDegrees % 360.0;
    if (aspectNorm > 180.0) {
      aspectNorm = 360.0 - aspectNorm;
    }

    // 1. Current state
    final currentOrb = getOrb(pointOneAbsPos, pointTwoAbsPos, aspectNorm);

    // 2. Future state
    // Note: % 360.0 in Dart behaves correctly for positive numbers.
    // If speed is negative and result is negative, we might need to adjust.
    // Dart's % operator returns result with same sign as dividend.
    // So -10 % 360 = -10. We want 350.

    double normalize(double v) {
      v = v % 360.0;
      if (v < 0) v += 360.0;
      return v;
    }

    final p1FutureNorm = normalize(pointOneAbsPos + pointOneSpeed * dt);
    final p2FutureNorm = normalize(pointTwoAbsPos + pointTwoSpeed * dt);

    final futureOrb = getOrb(p1FutureNorm, p2FutureNorm, aspectNorm);

    // 3. Compare
    final orbChange = futureOrb - currentOrb;

    if (orbChange.abs() < orbEpsilon) {
      return AspectMovementType.Static;
    } else if (orbChange < 0) {
      return AspectMovementType.Applying;
    } else {
      return AspectMovementType.Separating;
    }
  }

  /// Helper to get active points from subject
  static List<KerykeionPointModel> getActivePointsList(AstrologicalSubjectModel subject, List<AstrologicalPoint>? activePoints) {
    // If activePoints is null/empty, we should probably return DEFAULT points?
    // The Python implementation takes active_points arg and if None, returns empty list?
    // Wait, Python: "if active_points is None: active_points = []".
    // Then iterates over DEFAULT_CELESTIAL_POINTS_SETTINGS.
    // If "planet['name'] in active_points", it appends.

    // In Dart factory, we call it with `subject.activePoints` usually.

    final pointsToFilter = activePoints ?? subject.activePoints;
    final List<KerykeionPointModel> result = [];

    // Map through default settings to maintain order?
    // Or just iterate through defaults and check if present in subject?

    // The Python implementation uses `celestial_points` setting to order.
    // We should do the same.

    for (var setting in defaultCelestialPointsSettings) {
      // Find the enum from string name
      AstrologicalPoint? pointEnum;
      try {
        pointEnum = AstrologicalPoint.values.firstWhere((e) => e.toString().split('.').last == setting.name);
      } catch (e) {
        // Point name in settings might not match enum exactly or not exist
        continue;
      }

      if (pointsToFilter.contains(pointEnum)) {
        // Get the point from the subject
        // Subject properties are named camelCase (sun, moon, etc.)
        // We need a way to access properties dynamically or via a map.
        // AstrologicalSubjectModel extends SubscriptableBaseModel in Python.
        // In Dart, we might need a helper method in the model or here.

        final point = _getPointFromSubject(subject, pointEnum);
        if (point != null) {
          result.add(point);
        }
      }
    }

    return result;
  }

  static KerykeionPointModel? _getPointFromSubject(AstrologicalSubjectModel subject, AstrologicalPoint point) {
    switch (point) {
      case AstrologicalPoint.Sun:
        return subject.sun;
      case AstrologicalPoint.Moon:
        return subject.moon;
      case AstrologicalPoint.Mercury:
        return subject.mercury;
      case AstrologicalPoint.Venus:
        return subject.venus;
      case AstrologicalPoint.Mars:
        return subject.mars;
      case AstrologicalPoint.Jupiter:
        return subject.jupiter;
      case AstrologicalPoint.Saturn:
        return subject.saturn;
      case AstrologicalPoint.Uranus:
        return subject.uranus;
      case AstrologicalPoint.Neptune:
        return subject.neptune;
      case AstrologicalPoint.Pluto:
        return subject.pluto;
      case AstrologicalPoint.Mean_North_Lunar_Node:
        return subject.meanNorthLunarNode;
      case AstrologicalPoint.True_North_Lunar_Node:
        return subject.trueNorthLunarNode;
      case AstrologicalPoint.Mean_South_Lunar_Node:
        return subject.meanSouthLunarNode;
      case AstrologicalPoint.True_South_Lunar_Node:
        return subject.trueSouthLunarNode;
      case AstrologicalPoint.Chiron:
        return subject.chiron;
      case AstrologicalPoint.Mean_Lilith:
        return subject.meanLilith;
      case AstrologicalPoint.True_Lilith:
        return subject.trueLilith;
      case AstrologicalPoint.Earth:
        return subject.earth;
      case AstrologicalPoint.Pholus:
        return subject.pholus;
      case AstrologicalPoint.Ceres:
        return subject.ceres;
      case AstrologicalPoint.Pallas:
        return subject.pallas;
      case AstrologicalPoint.Juno:
        return subject.juno;
      case AstrologicalPoint.Vesta:
        return subject.vesta;
      case AstrologicalPoint.Eris:
        return subject.eris;
      case AstrologicalPoint.Sedna:
        return subject.sedna;
      case AstrologicalPoint.Haumea:
        return subject.haumea;
      case AstrologicalPoint.Makemake:
        return subject.makemake;
      case AstrologicalPoint.Ixion:
        return subject.ixion;
      case AstrologicalPoint.Orcus:
        return subject.orcus;
      case AstrologicalPoint.Quaoar:
        return subject.quaoar;
      case AstrologicalPoint.Regulus:
        return subject.regulus;
      case AstrologicalPoint.Spica:
        return subject.spica;
      case AstrologicalPoint.Pars_Fortunae:
        return subject.parsFortunae;
      case AstrologicalPoint.Pars_Spiritus:
        return subject.parsSpiritus;
      case AstrologicalPoint.Pars_Amoris:
        return subject.parsAmoris;
      case AstrologicalPoint.Pars_Fidei:
        return subject.parsFidei;
      case AstrologicalPoint.Vertex:
        return subject.vertex;
      case AstrologicalPoint.Anti_Vertex:
        return subject.antiVertex;
      case AstrologicalPoint.Ascendant:
        return subject.ascendant;
      case AstrologicalPoint.Medium_Coeli:
        return subject.mediumCoeli;
      case AstrologicalPoint.Descendant:
        return subject.descendant;
      case AstrologicalPoint.Imum_Coeli:
        return subject.imumCoeli;
    }
  }

  static int getPlanetId(String name) {
    final setting = defaultCelestialPointsSettings.firstWhere(
      (element) => element.name == name,
      orElse: () => CelestialPointSetting(id: -1, name: name, color: "", elementPoints: 0, label: ""),
    );
    return setting.id;
  }
}
